server/src/main/java/org/apache/iotdb/db/query/fill/PreviousFill.java

```diff
@@ -59,28 +84,183 @@ public long getBeforeRange() {
+    PriorityQueue<TsFileResource> unseqFileResource =
+        sortUnSeqFileResourcesInDecendingOrder(dataSource.getUnseqResources());
+
+    while (!unseqFileResource.isEmpty()) {
+      // The very end time of unseq files is smaller than lBoundTime,
+      // then skip all the rest unseq files
+      if (unseqFileResource.peek().getEndTimeMap().get(seriesPath.getDevice()) < lBoundTime) {
+        return;
+      }
+      TimeseriesMetadata timeseriesMetadata =
+          FileLoaderUtils.loadTimeSeriesMetadata(
+              unseqFileResource.poll(), seriesPath, context, timeFilter, allSensors);
+      if (timeseriesMetadata != null
+          && lBoundTime <= timeseriesMetadata.getStatistics().getEndTime()) {
+        lBoundTime = Math.max(lBoundTime, timeseriesMetadata.getStatistics().getStartTime());
+        unseqTimeseriesMetadataList.add(timeseriesMetadata);
         break;
       }
     }
 
-    if (beforePair != null) {
-      beforePair.setTimestamp(queryTime);
-    } else {
-      beforePair = new TimeValuePair(queryTime, null);
+    // unpack all overlapped unseq files and fill unseqTimeseriesMetadata list
+    while (!unseqFileResource.isEmpty()
+        && (lBoundTime <= unseqFileResource.peek().getEndTimeMap().get(seriesPath.getDevice()))) {
+      TimeseriesMetadata timeseriesMetadata =
+          FileLoaderUtils.loadTimeSeriesMetadata(
+              unseqFileResource.poll(), seriesPath, context, timeFilter, allSensors);
+      unseqTimeseriesMetadataList.add(timeseriesMetadata);
+      // current unseq timeseriesMetadata's last point is a valid result,
+      // then skip the rest unseq files
+      if (endtimeContainedByTimeFilter(timeseriesMetadata.getStatistics())
+          && timeseriesMetadata.getStatistics().getEndTime()
+              > unseqFileResource.peek().getEndTimeMap().get(seriesPath.getDevice())) {
+        break;
+      }
     }
-    return beforePair;
+  }
+
+  private TimeValuePair getChunkLastPoint(ChunkMetadata chunkMetaData) throws IOException {
+    TimeValuePair lastPoint = new TimeValuePair(Long.MIN_VALUE, null);
+    Statistics chunkStatistics = chunkMetaData.getStatistics();
+
+    if (endtimeContainedByTimeFilter(chunkStatistics)) {
+      return constructLastPair(
+          chunkStatistics.getEndTime(), chunkStatistics.getLastValue(), dataType);
+    }
+    List<IPageReader> pageReaders = FileLoaderUtils.loadPageReaderList(chunkMetaData, timeFilter);
+    for (int i = pageReaders.size() - 1; i >= 0; i--) {
+      IPageReader pageReader = pageReaders.get(i);
+      Statistics pageStatistics = pageReader.getStatistics();
+      if (endtimeContainedByTimeFilter(pageStatistics)) {
+        lastPoint = constructLastPair(
+            pageStatistics.getEndTime(), pageStatistics.getLastValue(), dataType);
+      } else {
+        BatchData batchData = pageReader.getAllSatisfiedPageData();
+        lastPoint = batchData.getLastPairBeforeOrEqualTimestamp(queryTime);
+      }
+      break;
+    }
```

Comment:
with a `break`, the for loop is not used

Reply:
Fixed